Use HTTPS for Salt API

slaperche-scality opened the pull request:
**Component**:

salt

**Context**: 

We want to use HTTPS to ensure secrecy of authentication, as well as data transmitted between Salt and the clients.

**Summary**:

- generate a certificate for Salt-API
- update the bootstrap script to generate the root CA before deploying the Salt master (because we need the root CA to sign the Salt API certs)
- fix Salt states to use HTTPS
- fix probes to use HTTPS
- fix the UI to use HTTPS
- fix the storage operator to use HTTPS
- update the documentation accordingly

**Acceptance criteria**: 

Salt-API use HTTPS.
Storage-Operator still works.
Platform UI still works.

---

Closes: #1115

slaperche-scality committed:
salt/certs: generate certificates for Salt API

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
salt/salt-api: update the config to use HTTPS

Also update the salt-api container manifest to:
- mount the certs
- use HTTPs for readiness and liveness probes

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Hello slaperche-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

TeddyAndrieux commented on salt/metalk8s/salt/master/certs/salt-api.sls:
```diff
@@ -0,0 +1,33 @@
+    - name: /etc/kubernetes/pki/salt-api.crt
+    - public_key: /etc/kubernetes/pki/salt-api.key
+    - ca_server: {{ pillar['metalk8s']['ca']['minion'] }}
+    - signing_policy: {{ etcd.cert.apiserver_client_signing_policy }}
```
You should use this signing policy ` kube_api.cert.server_signing_policy` as it's a server auth certs IMO

TeddyAndrieux replied:
You also need to add some subjectAltName or change the `CN`.
For me we want at least our control_plane ip and fqdn, hostname (maybe `localhost/127.0.0.1` but... maybe not) as subjectAltName.
Maybe @NicolasT has a precise idea for this one

slaperche-scality replied:
Meh, too much copy-pasta :sweat_smile:

NicolasT replied:
As @TeddyAndrieux already mentioned, indeed, this should be a server policy being applied, otherwise the cert won't even work.

As for the CN, let's not simply use `salt-api`, but a more descriptive name, e.g. `salt-api on {{ grains.id }}` or something along those lines.

Then, we indeed should set up a proper list of `subjectAltName`s. This should include
- For *external* access:
  * The control-plane IP of the host, `grains['metalk8s']['control_plane_ip']`
  * `grains['fqdn']`
- For *internal* access: the DNS names of the Salt service (when running in the cluster), with various suffices:
  * `salt-master`
  * `salt-master.kube-system`
  * `salt-master.kube-system.svc`
  * `salt-master.kube-system.svc.cluster.local`

slaperche-scality replied:
Thanks for the info.

Done.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @teddyandrieux

slaperche-scality committed:
salt: split CA generation and export

Until now, the state managing the Kubernetes API server CA was doing two
things at once:
- generate the key and certificate
- pushing them into the Salt mine

Now, we split those two step into two different steps, because we need
to:
- generate the Kubernetes API server CA
- deploying the salt-master (which will need the Kubernetes API server
CA to sign the Salt API one)
- push them into the mine

Note that we may want to also do this for the other states that manage
certificates, but that's out of the scope of this commit/issue.

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT commented on scripts/bootstrap.sh.in:
```diff
@@ -261,7 +261,7 @@ orchestrate_bootstrap() {
     run "Deploying early-stage bootstrap node in local mode (this may take a while)" \
         "${SALT_CALL}" --local --state-output=mixed --retcode-passthrough state.sls \
-            '["metalk8s.roles.bootstrap", "metalk8s.roles.minion"]' \
+            '["metalk8s.kubernetes.ca.kubernetes.installed", "metalk8s.roles.bootstrap", "metalk8s.roles.minion"]' \
```
I think it makes sense to only keep `metalk8s.roles.*` SLSs here.

Instead, couldn't the state in which we create a config file which points to this key/cert pair 'include' the relevant state?

TeddyAndrieux replied:
The problem is that we generate these key/cert here because we need them to create the salt-api certificate but we not use the CA key/cert 
If we want to include the state somewhere it will likely be in the salt-master state or in the roles.bootstrap state.
Personnaly don't really like it, that mean bootstrap roles will also deploy the kubernetes root CA. Which is not what we want IMO

NicolasT replied:
Well, we *do* use that CA's key/cert, *exactly* to generate the SaltAPI cert.

- `metalk8s.roles.bootstrap` requires the CA to be available, otherwise we can't sign the SaltAPI cert it wants/needs to deploy
- Implicitly, this means `metalk8s.roles.bootstrap` (or a state it contains, e.g. `salt-master`) requires the `ca` minion to be provisioned
- We're in bootstrap phase of the cluster, so `ca` minion == bootstrap minion (there's only one node...)
- Ergo, how about simply adding `metalk8s.roles.ca` in this list, before `metalk8s.roles.bootstrap`?

TeddyAndrieux replied:
> * Ergo, how about simply adding `metalk8s.roles.ca` in this list, before `metalk8s.roles.bootstrap`?

Yes we can but we need to rewrite all the CA states because for the moment Ca states also push the certificate to the mine but if we have no salt-master we can't push to the mine. (At least that does not fail but nothing is really pushed, so maybe we can do that but we will need to run the state again after the salt-master deployment)

> * `metalk8s.roles.bootstrap` requires the CA to be available, otherwise we can't sign the SaltAPI cert it wants/needs to deploy
> * Implicitly, this means `metalk8s.roles.bootstrap` (or a state it contains, e.g. `salt-master`) requires the `ca` minion to be provisioned

Yes I agree, but the salt states are not used only for the first deployment compare to the bootstrap script so for me it's better to have this behavior only in the script so that we don't have a requirement in our salt formulas to use the same node for `ca` and `salt-master`

NicolasT replied:
OK

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/1115-use-https-for-salt-api` and the
destination branch `development/2.4`.

Please resolve the conflict on **the feature branch** (`improvement/1115-use-https-for-salt-api`).

```bash
 $ git fetch
 $ git checkout origin/improvement/1115-use-https-for-salt-api
 $ git merge origin/development/2.4
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/1115-use-https-for-salt-api
```

slaperche-scality committed:
salt/ui: use HTTPS for Salt API

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
Bootstrap now working, still have to fix the storage-operator though…
And the UI also have issue to connect…

slaperche-scality committed:
doc/developer: update the script dumping the UI conf

We need to prefix the IP by `https` instead of `http`.

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality committed:
ui/login: add a link to accept the Salt API cert

Otherwise we can't move past the login page because the browser won't
allow the requests to Salt API.

Update the documentation accordingly.

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
UI fixed, no let's tackle the storage operator.

slaperche-scality committed:
storage-operator: use HTTPS to communicate with the Salt API

Note that we have some logic to fetch the root CA cert used to validate
the cert provided by Salt API.

Indeed, the `rest.Config`[1] object does contains an inlined
`TLSClientConfig` structure that contains the following fields[2]:
- `CAFile`: the path to the file that contains the root certificate
- `CAData`: the PEM-encoded root certificates (takes precedence over the
`CAFile` field).

From what I've observed:
- when the operator runs inside the cluster `CAData` is empty, we have
to read the certificate from `CAFile`.
- when the operator runs outside of the cluster, `CAData` is not empty
and can be used.

[1]: https://godoc.org/k8s.io/client-go/rest#Config
[2]: https://godoc.org/k8s.io/client-go/rest#TLSClientConfig

Closes: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

slaperche-scality commented:
With that one, everything should be OK, let's see if it pass the tests…

slaperche-scality committed:
test/salt: use HTTPS for Salt API auth

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

NicolasT reviewed (commented):
LGTM so far

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

* @teddyandrieux

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality





The following reviewers are expecting changes from the author, or must review again:

* @nicolast

charlesprost commented on salt/metalk8s/kubernetes/ca/kubernetes/init.sls:
```diff
@@ -5,7 +5,8 @@
 # ================
 #
 # * installed   -> install and advertise as CA server
```
```suggestion
# * installed   -> install as CA server
```
This `installed` state is not advertising anymore.

slaperche-scality replied:
Done

charlesprost commented on salt/metalk8s/salt/master/certs/salt-api.sls:
```diff
@@ -0,0 +1,46 @@
+    - verbose: False
+    - user: root
+    - group: root
+    - mode: 600
```
Why not 400? even if that does not change a lot...

charlesprost replied:
Or is that because if we want to renew it we would need 600?

slaperche-scality replied:
Good question.
But since it's like that for every certs, if we want to change it we should probably change it for all certs

charlesprost commented on salt/metalk8s/salt/master/certs/salt-api.sls:
```diff
@@ -0,0 +1,46 @@
+
+Create Salt API private key:
+  x509.private_key_managed:
+    - name: /etc/kubernetes/pki/salt-api.key
```
Salt and K8s are quite two different worlds. Should not we have a different path for both? Like `/etc/salt-api/pki/`? or something within `/etc/metalk8s`?

NicolasT replied:
Good catch @nootal. This key/cert should indeed go in `/etc/salt/pki/api` (one could argue for `/etc/salt/pki/master` somewhere as well, but there's a reason not to, read on). This way, there's no need to expose anything SaltAPI *doesn't* need (e.g., access to K8s API server certs, Salt master certs,...) to its container.

slaperche-scality replied:
Good point indeed.

That being said, I still need to mount the K8S stuff, to access the root CA, no?

slaperche-scality replied:
Done in a new commit.

charlesprost commented on salt/metalk8s/salt/master/certs/salt-api.sls:
```diff
@@ -0,0 +1,46 @@
+    - dir_mode: 755
+    - require:
+      - x509: Create Salt API private key
+
```
Trailing

slaperche-scality replied:
Done

charlesprost commented on salt/metalk8s/salt/master/files/master-99-metalk8s.conf.j2:
```diff
@@ -25,7 +25,9 @@ no_host_keys: true
   port: 4507
   host: {{ salt_ip }}
-  disable_ssl: true
+  ssl_crt: /etc/kubernetes/pki/salt-api.crt
```
Should not we put those paths into some pillars or something to make them defined at one location only?

NicolasT replied:
No, because you *choose* where the host path is exposed into a container, and here we refer to the in-container path.

charlesprost commented on salt/metalk8s/salt/master/installed.sls:
```diff
@@ -57,7 +57,8 @@ Make sure salt master container is up:
 Wait for Salt API to answer:
   http.wait_for_successful_query:
-    - name: http://{{ salt_ip }}:4507/
+    - name: https://{{ salt_ip }}:4507/
```
There is no convention about changing the port for https? Just asking

slaperche-scality replied:
Not that I know (even the 4507 is not a "standard" Salt API port)

charlesprost reviewed (commented):
Well done!

slaperche-scality committed:
salt/certs: move Salt API certs under /etc/salt

It's more about Salt than K8S so it's better to put it here.

Refs: #1115
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

slaperche-scality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye slaperche-scality.
